policy/policy.go

```diff
@@ -0,0 +1,406 @@
+		}
+	}
+	if len(errStr) > 0 {
+		return common.NewError500(errStr)
+	}
+	return nil
+}
+
+func (policy *PolicySvc) getPolicy(input interface{}, ctx common.RestContext) (interface{}, error) {
+	idStr := ctx.PathVariables["policyID"]
+	id, err := strconv.ParseUint(idStr, 10, 64)
+	if err != nil {
+		return nil, common.NewError404("policy", idStr)
+	}
+	policyDoc, err := policy.store.getPolicy(id, false)
+	log.Printf("Found policy for ID %d: %s (%v)", id, policyDoc, err)
+	return policyDoc, err
+}
+
+func (policy *PolicySvc) deletePolicyHandler(input interface{}, ctx common.RestContext) (interface{}, error) {
+	idStr := strings.TrimSpace(ctx.PathVariables["policyID"])
+	if idStr == "" {
+		if input == nil {
+			return nil, common.NewError400("Request must either be to /policies/{policyID} or have a body.")
+		}
+		policyDoc := input.(*common.Policy)
+		err := policyDoc.Validate()
+		if err != nil {
+			return nil, err
+		}
+		id, err := policy.store.lookupPolicy(policyDoc.ExternalID, policyDoc.Datacenter.Id)
+		log.Printf("Found %d / %v (%T) from external ID %s", id, err, err, policyDoc.ExternalID)
+		if err != nil {
+			return nil, err
+		}
+		return policy.deletePolicy(id)
+	} else {
+		if input != nil {
+			common.NewError400("Request must either be to /policies/{policyID} or have a body.")
+		}
+		id, err := strconv.ParseUint(idStr, 10, 64)
+		if err != nil {
+			return nil, common.NewError404("policy", idStr)
+		}
+		return policy.deletePolicy(id)
+	}
+}
+
+// deletePolicy deletes policy based the following algorithm:
+//1. Mark the policy as "deleted" in the backend store.
+func (policy *PolicySvc) deletePolicy(id uint64) (interface{}, error) {
+	// TODO do we need this to be transactional or not ... case can be made for either.
+	err := policy.store.inactivatePolicy(id)
+	if err != nil {
+		return nil, err
+	}
+	policyDoc, err := policy.store.getPolicy(id, true)
+	log.Printf("Found policy for ID %d: %s (%v)", id, policyDoc, err)
+	if err != nil {
+		return nil, err
+	}
+	hosts, err := policy.client.ListHosts()
+	if err != nil {
+		return nil, err
+	}
+	errStr := make([]string, 0)
+	for _, host := range hosts {
+		// TODO make schema configurable
+		url := fmt.Sprintf("http://%s:%d/policies", host.Ip, host.AgentPort)
+		result := make(map[string]interface{})
+		err = policy.client.Delete(url, policyDoc, result)
+		log.Printf("Agent at %s returned %v", host.Ip, result)
+		if err != nil {
+			errStr = append(errStr, fmt.Sprintf("Error deleting policy %d (%s) from host %s: %v. ", id, policyDoc.Name, host.Ip, err))
+		}
+	}
+	if len(errStr) > 0 {
+		return nil, common.NewError500(errStr)
+	}
+	err = policy.store.deletePolicy(id)
+	if err != nil {
+		return nil, err
+	}
+	return policyDoc, nil
+}
+
+// deletePolicy deletes policy...
+func (policy *PolicySvc) listPolicies(input interface{}, ctx common.RestContext) (interface{}, error) {
+	return policy.store.listPolicies()
+}
+
+// findPolicyByName returns the first policy found corresponding
```

Comment:
Why does it not return all policies with a given name? Since policies can be defined by tenants, should there not be a tenant specified as well? If only the first one is returned, but we have 20 tenants, 10 of which defined a policy with the same name, how useful is this?